AISDK-96: Add javadoc notation and populate README (#13)
revdotcom/revai-java-sdk: README.md

[![Build Status](https://travis-ci.org/revdotcom/revai-java-sdk.svg?branch=develop)](https://travis-ci.org/revdotcom/revai-java-sdk)

## Documentation

See the [API docs](https://www.ai.rev.ai/docs) for more information about the API.

## Install the SDK
The recommended way to use the Rev.ai Java SDK is to import it into the project using Maven.

      <dependency>

>         <groupId>ai.rev</groupId>

Reviewer: The reverse domain format is a bit confusing with short domain names and non-dotcom domain suffixes.

As I understand it, this is a requirement for our groupId to match our domain name. Is this correct?

Author: Yes and this follows maven groupid naming convention.
http://maven.apache.org/guides/mini/guide-naming-conventions.html

        <artifactId>revai-java-sdk</artifactId>
        <version>1.0.0</version>
      </dependency>
    
## Installing it locally
Once you've cloned the repo you can use Maven to build it locally and install it your local Maven .m2 repository.

    mvn install -Dmaven.test.skip=true

## Support

We support Java 8 and 11.

## Usage

All you need to get started is your Access Token, which can be generated on
your [Settings Page](https://www.ai.rev.ai/access_token). Create a client with the
given Access Token:

```
// Initialize your client with your ai.rev access token
String accessToken = "Your Access Token";
ApiClient apiClient = new ApiClient(accessToken);
```

### Checking credits remaining

```
RevAiAccount revAiAccount = apiClient.getAccount(); 
```

### Submitting a job

You can submit a local file
```
RevAiJob revAiJob = apiClient.submitJobLocalFile("./path/to/file.mp3");
```

or submit via a public url
```

> RevAiJob revAiJob = apiClien.submitJobUrl("https://www.ai.ai.rev.ai/FTC_Sample_1.mp3");

Reviewer: ```suggestion
RevAiJob revAiJob = apiClient.submitJobUrl("https://www.ai.ai.rev.ai/FTC_Sample_1.mp3");
```

```

or from FileInputStream, the filename is optional.
```
File file = new File("./path/to/file.mp3");
FileInputStream fileInputStream;
try {
  fileInputStream = new FileInputStream(file);
} catch (FileNotFoundException e) {
  throw new RuntimeException("Could not find file [" + file.getName() + "]");
}
RevAiJob revAiJob = apiClient.submitJobLocalFile(fileInputStream, String fileName, RevAiJobOptions options);
```

> `RevAiJob` objects contain all the information normally found in a successful response from our

Reviewer: ```suggestion
`RevAiJob` objects contain job information as defined by the documentation...
```

Reviewer: I think you can omit the concept of success or failure here. We have a separate section for Errors.

[Submit Job](https://www.ai.rev.ai/docs#operation/SubmitTranscriptionJob) endpoint.

> If you want to get fancy, all send job methods can take a `RevAiJobOptions` object which contains methods 

Reviewer: Use consistent terminology:
```suggestion
If you want to get fancy, all submit job methods have overrides that allow specifying `RevAiJobOptions` to configure job specific settings: `metadata`, ...
```

Reviewer: I would also mention the list is non-exhaustive or reference a list of job options or the job options object definition instead of listing them out. We don't want to have to update this README's list every time we add a new job option.

to set optional parameters such as: `metadata`, `callback_url`, `skip_diarization`,`skip_punctuation`, 
`speaker_channels_count`, `filter_profanity`, `remove_disfluencies`  and `custom_vocabularies`. These are 
also described in the request body of the [Submit Job](https://www.ai.rev.ai/docs#operation/SubmitTranscriptionJob) endpoint.

### Checking your job's status

You can check the status of your transcription job using its `id`

```

> RevAiJob revAiJob = apiClient.getJobDetails(revAiJob.getJobID());

Reviewer: This example contains a cyclical reference. Without context it doesn't make complete sense.

You could change the resulting revAiJob's name to `newlyRefreshedRevAiJob`

```

> `RevAiJob` objects contain all information normally found in a successful response from

Reviewer: Duplicate information from the previous. I would recommend my wording that I mention above.

our [Get Job](https://www.ai.rev.ai/docs#operation/GetJobById) endpoint

### Checking multiple files

You can retrieve a list of transcription jobs with optional parameters

```

> List<RevAiJob> jobs = apiClient.getListOfJobs();

Reviewer: Minor -  why didn't we call the method name simply`getJobs()`?

Author: It's not too late 😉

Author: But this does match the documentation https://www.rev.ai/docs#operation/GetListOfJobs

Reviewer: I'm not a fan of that wording in our documentation. I think GET /jobs implies that it will return a list of some kind.

Reviewer: I don't want to change the wording if we use the same `getlistofjobs` in other sdks. The higher priority is being consistent.

Author: Yes this is consistent with the other SDKs as well

// limit amount of retrieved jobs

> List<RevAiJob> jobs = apiClient.getListOfJobs(3);

Reviewer: use local variables to ensure clarity since java does not allow named parameters.

// get jobs starting after a certain job id

> List<RevAiJob> jobs = apiClient.getListOfJobs("Umx5c6F7pH7r");

Reviewer: ditto on local variable

[![Build Status](https://travis-ci.org/revdotcom/revai-java-sdk.svg?branch=develop)](https://travis-ci.org/revdotcom/revai-java-sdk)

## Documentation

See the [API docs](https://www.ai.rev.ai/docs) for more information about the API.

## Install the SDK
The recommended way to use the Rev.ai Java SDK is to import it into the project using Maven.

      <dependency>
        <groupId>ai.rev</groupId>
        <artifactId>revai-java-sdk</artifactId>
        <version>1.0.0</version>
      </dependency>
    
## Installing it locally

> Once you've cloned the repo you can use Maven to build it locally and install it your local Maven .m2 repository.

Reviewer 2: install it `in` your local Maven ...?

    mvn install -Dmaven.test.skip=true

## Support

We support Java 8 and 11.

## Usage

All you need to get started is your Access Token, which can be generated on
your [Settings Page](https://www.ai.rev.ai/access_token). Create a client with the
given Access Token:

```
// Initialize your client with your ai.rev access token
String accessToken = "Your Access Token";
ApiClient apiClient = new ApiClient(accessToken);
```

### Checking credits remaining

```
RevAiAccount revAiAccount = apiClient.getAccount(); 
```

### Submitting a job

You can submit a local file
```
RevAiJob revAiJob = apiClient.submitJobLocalFile("./path/to/file.mp3");
```

or submit via a public url
```

> RevAiJob revAiJob = apiClient.submitJobUrl("https://www.ai.ai.rev.ai/FTC_Sample_1.mp3");

Reviewer 2: is that a real media url?

Author: I used to be lol. It got messed up in a refactor. I thought I found all of those. Good catch!

[![Build Status](https://travis-ci.org/revdotcom/revai-java-sdk.svg?branch=develop)](https://travis-ci.org/revdotcom/revai-java-sdk)

## Documentation

See the [API docs](https://www.ai.rev.ai/docs) for more information about the API.

## Install the SDK
The recommended way to use the Rev.ai Java SDK is to import it into the project using Maven.

      <dependency>
        <groupId>ai.rev</groupId>
        <artifactId>revai-java-sdk</artifactId>
        <version>1.0.0</version>
      </dependency>
    

> ## Installing it locally

Reviewer 3: Maybe we should call this section "Build and install locally from source". Otherwise it's a bit unclear how it relates to the previous section (install the SDK).

```

or submit via a public direct download url
```
String urlLinkToFile = "https://www.rev.ai/FTC_Sample_1.mp3";
RevAiJob revAiJob = apiClient.submitJobUrl(urlLinkToFile);
```

or from FileInputStream, the filename is optional.
```
File file = new File("./path/to/file.mp3");
FileInputStream fileInputStream;
try {
  fileInputStream = new FileInputStream(file);
} catch (FileNotFoundException e) {
  throw new RuntimeException("Could not find file [" + file.getName() + "]");
}
RevAiJob revAiJob = apiClient.submitJobLocalFile(fileInputStream, String fileName, RevAiJobOptions options);
```

`RevAiJob` objects contain job information as defined by the [documentation](https://www.ai.rev.ai/docs#operation/SubmitTranscriptionJob).

If you want to get fancy, all submit job methods have overrides that allow specifying 
[RevAiJobOptions](./src/main/java/ai/rev/models/asynchronous/RevAiJobOptions.java) to configure job specific settings. These are also described in the 
request body of the [Submit Job](https://www.ai.rev.ai/docs#operation/SubmitTranscriptionJob) 
endpoint.

### Checking your job's status

You can check the status of your transcription job using its `id`

```

> RevAiJob newlyRefreshedRevAiJob = apiClient.getJobDetails(revAiJob.getJobID());

Reviewer 3: small nitpick on the theme of naming, but could we use `JobId` instead of `JobID`everywhere (and in general Id vs ID). Generally we use Id in all of our other code.

Author: Good catch!

    }

    @Override
    public void onError(Throwable t, Response response) {
        System.out.println("On Error: " + response.toString());
    }

    @Override
    public void onClose(int code, String reason) {
        System.out.println("On Close: [" + code + "] " + reason);
    }

    @Override
    public void onOpen(Response response) {
        System.out.println("On Open: " + response.toString());
    }
}
```

Now you will be able to connect and start the streaming session by calling the `streamingClient.connect()` method and passing in the Listener! You can supply an optional `SessionConfig` object to the function in order to provide additional information for that session, such as metadata or enabled the profanity filter to be used.

```
Listener listener = new Listener();

SessionConfig sessionConfig = new SessionConfig();
sessionConfig.setMetaData("My first job");
sessionConfig.setFilterProfanity(true);

streamingClient.connect(clientListener, streamContentType, sessionConfig);
```

You can stream data over the WebSocket in the form of a `ByteString` using the `streamingClient.sendAudioData()` method.

```
streamingClient.sendAudioData(ByteString);
```

> The streaming connection will close when you call the method `streamingClient.close()` or if you go 15 seconds without sending any audio data.

Reviewer 3: It would be good to provide guidance on how a StreamingClient instance should be reused (or not) across multiple streaming sessions (both simultaneous and not).
